staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -671,33 +725,42 @@ func (c *Cacher) processEvent(event *watchCacheEvent) {
 }
 
 func (c *Cacher) dispatchEvents() {
+	startBucketID := c.bookmarkWatchers.currentBucketID()
 	for {
 		select {
 		case event, ok := <-c.incoming:
 			if !ok {
 				return
 			}
-			c.dispatchEvent(&event)
+			endBucketID := c.bookmarkWatchers.currentBucketID() + 1
+			startBucketID = c.dispatchEvent(&event, startBucketID, endBucketID)
```

Comment:
Instead of passing those bucketIDs back and forth, it would be much easier to add a "currentBucketID" field (or sth like that) field as a cacherField - given that you only use it under lock, it would be much cleaner.